src/gasman.c

```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
 }
 
 
-/****************************************************************************
-**
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// As ChangedBags is a singly linked list, we can't easily take anything out of
+// this list. Also, as the link pointers are stored inside the bag itself, when
+// we swap PTR_BAG, we must swap LINK_BAG as well.
+//
+// If both bags are old, and the original bag1 was changed, we need to now mark
+// bag2 changed (and vice-versa).
+//
+// If both bags are young, GAP assumes they are both changed.
```

Comment:
It seems my previous comments are now all hidden, due to the revised commit, yet they were not all addressed :-(.

In particular, I still don't understand what the purpose of this sentence is.... Anyway, all in all, how about this after "For completeness and future reference ...":

"When swapping two master pointers, we have to take into account whether they are marked as changed or not, as we otherwise may end up in an inconsistent state, were some object is referenced by another, but GASMAN is lead to believe this is not so; GASMAN then collects the bag, resulting in a corrupted workspace.

To understand this better, we consider the following three cases:

1. Both bags are old. Then if the original bag1 had been previously marked as changed (by having been put into the ChangedBags simply linked list), then we must make sure to mark the new bag2 as changed, too (and vice-versa).

2. Both bags are young. Then they typically will not be on the list of changed bags (GAP implicitly treats all young bags as if they had been marked changed), so nothing needs to be done. (However, while CHANGED_BAG will indeed never put a young bag on the list of changed bags, this can still happen in step 3, so strictly speaking we need to do something similar in step 2 as in step 1).

3. Bag1 is young and bag2 is old (or vice-versa). Then [...]
"
now continue with the existing comment.